cabal-install/src/Distribution/Client/CmdStatus.hs

```diff
@@ -0,0 +1,381 @@
+      compilerProg <- requireCompilerProg verbosity compiler
+      let progDb = pkgConfigCompilerProgs elabSharedConfig
+      (configuredCompilerProg, _) <- requireProgram verbosity compilerProg progDb
+      pure $ Just $ mkCompilerInfo configuredCompilerProg compiler
+
+  resolvedTargets <- if null (statusTargets statusFlags)
+    then pure Nothing
+    else do
+      let targetStrings = statusTargets statusFlags
+      mtargetSelectors <- mapM (readTargetSelector (localPackages baseCtx) Nothing) targetStrings
+      let (unresolvable, targetSelectors) = partitionEithers
+                    $ map (\(mts, str) -> case mts of
+                        Left _ -> Left str
+                        Right ts -> Right (ts, str)
+                    )
+                    $ zip mtargetSelectors targetStrings
+      -- Interpret the targets on the command line as build targets
+      -- (as opposed to say repl or haddock targets).
+      -- TODO: don't throw on targets that are invalid.
+      -- TODO: why might this still fail? should we try to avoid that?
+      targets <- either (reportBuildTargetProblems verbosity) return
+                $ resolveTargets
+                    selectPackageTargets
+                    selectComponentTarget
+                    elaboratedPlan
+                    Nothing
+                    (map fst targetSelectors)
+
+      pure $ Just $ mkBuildInfoJson elaboratedPlan targets (Map.fromList targetSelectors) unresolvable
+
+  let si = StatusInformation
+        { siCabalVersion = cabalInstallVersion
+        , siCompiler = compilerInformation
+        , siTargetResolving = resolvedTargets
+        }
+
+  serialisedStatusInformation <- serialise verbosity (statusOutputFormat statusFlags) si
+
+  -- Final output
+  putStrLn $ withOutputMarker verbosity serialisedStatusInformation
+  where
+    verbosity = fromFlagOrDefault normal (configVerbosity configFlags)
+    cliConfig = commandLineFlagsToProjectConfig globalFlags flags mempty
+
+-- ----------------------------------------------------------------------------
+-- Big Datatype that can be serialised to different formats
+-- ----------------------------------------------------------------------------
+
+data StatusInformation = StatusInformation
+  { siCabalVersion :: Version
+  , siCompiler :: Maybe CompilerInformation
+  , siTargetResolving :: Maybe [ResolvedTarget]
+  }
+  deriving (Show, Read, Eq, Ord)
+
+data CompilerInformation = CompilerInformation
+  { ciFlavour :: CompilerFlavor
+  , ciCompilerId :: CompilerId
+  , ciPath :: FilePath
+  }
+  deriving (Show, Read, Eq, Ord)
+
+data ResolvedTarget = ResolvedTarget
+  { rtOriginalTarget :: String
+  -- | UnitId of the resolved target.
+  -- If 'Nothing', then the given target can not be resolved
+  -- to a target in this project.
+  , rtUnitId :: Maybe UnitId
+  }
+  deriving (Show, Read, Eq, Ord)
+
+serialise :: Verbosity -> Flag StatusOutputFormat -> StatusInformation -> IO String
```

Comment:
I'd suggest to return `Either String String` or use `MonadThrow` and turn it into `die` at the call site. Serialization shouldn't require IO.

Reply:
Resorted to `MonadError` which should have a similar effect.